File: FiniteStateMachine.Core/FiniteStateMachine.cs
namespace FiniteStateMachine.Core
{
    /// <summary>
    /// A generic finite state machine implementation.
    /// </summary>
    /// <typeparam name="TInput">The type of input that triggers state transitions.</typeparam>
    /// <typeparam name="TOutput">The type of output associated with states.</typeparam>
    public class FiniteStateMachine<TInput, TOutput>
    {
        private IState<TInput, TOutput> currentState;

        /// <summary>
        /// Creates a new instance of FiniteStateMachine.
        /// </summary>
        /// <param name="initialState">The initial state of the machine.</param>
        public FiniteStateMachine(IState<TInput, TOutput> initialState)
        {
            currentState = initialState;
        }

        /// <summary>
        /// Processes a sequence of inputs and returns the final state's output.
        /// </summary>
        /// <param name="inputs">The sequence of inputs to process.</param>
        /// <returns>The output of the final state.</returns>
        public TOutput ProcessInputs(IEnumerable<TInput> inputs)
        {
            foreach (var input in inputs)
            {
                currentState = currentState.ProcessInput(input);
            }
            return currentState.Output;
        }

        /// <summary>
        /// Gets the current state's output.
        /// </summary>
        public TOutput CurrentOutput => currentState.Output;

        /// <summary>
        /// Resets the machine to its initial state.
        /// </summary>
        /// <param name="initialState">The state to reset to.</param>
        public void Reset(IState<TInput, TOutput> initialState)
        {
            currentState = initialState;
        }
    }
}

File: FiniteStateMachine.Python/fsm/finite_state_machine.py
from typing import TypeVar, Generic, Iterable
from .state import State

TInput = TypeVar('TInput')
TOutput = TypeVar('TOutput')

class FiniteStateMachine(Generic[TInput, TOutput]):
    """A generic finite state machine implementation."""
    
    def __init__(self, initial_state: State[TInput, TOutput]):
        """Create a new instance of FiniteStateMachine."""
        self._current_state = initial_state
    
    def process_inputs(self, inputs: Iterable[TInput]) -> TOutput:
        """Process a sequence of inputs and return the final state's output."""
        for input_value in inputs:
            self._current_state = self._current_state.process_input(input_value)
        return self._current_state.output
    
    @property
    def current_output(self) -> TOutput:
        """Get the current state's output."""
        return self._current_state.output
    
    def reset(self, initial_state: State[TInput, TOutput]) -> None:
        """Reset the machine to its initial state."""
        self._current_state = initial_state
